### PR TITLE
fix: locked users admin list and bypass query

### PR DESCRIPTION
**Pull request by zomars**
## What does this PR do?

refs #17889

- Fixes bug where search on admin users list was not working.
- Fixed bug where  locked users disappear from listing.
- Refactors so we can bypass locked users on queries using AND and OR conditions.

This pull request includes changes to improve the handling of locked users in the Prisma extensions and the TRPC server. The most important changes include adding a utility function for safe JSON stringification, modifying the logic to exclude locked users, and updating search filters to bypass the exclusion of locked users.

Improvements to Prisma extensions:

* [`packages/prisma/extensions/exclude-locked-users.ts`](diffhunk://#diff-4e9949cb57d33c7ce0aa741636798e14df4ec4a352d412267a22a1ef5deb6a0eR28-R35): Added a new `safeJSONStringify` function to safely convert objects to JSON strings, and updated the `excludeLockedUsers` function to use this utility for checking if locked users should be included. [[1]](diffhunk://#diff-4e9949cb57d33c7ce0aa741636798e14df4ec4a352d412267a22a1ef5deb6a0eR28-R35) [[2]](diffhunk://#diff-4e9949cb57d33c7ce0aa741636798e14df4ec4a352d412267a22a1ef5deb6a0eR46-R49)

Updates to TRPC server:

* [`packages/trpc/server/routers/viewer/admin/listPaginated.handler.ts`](diffhunk://#diff-7a73fd2ab081d9ecb3457a7c1ddff8487f88251d118cbb5a5130184385a65f8cR20-R25): Introduced a `bothLockedAndUnlockedWhere` filter to bypass the `excludeLockedUsersExtension`, ensuring that both locked and unlocked users are included in search results. [[1]](diffhunk://#diff-7a73fd2ab081d9ecb3457a7c1ddff8487f88251d118cbb5a5130184385a65f8cR20-R25) [[2]](diffhunk://#diff-7a73fd2ab081d9ecb3457a7c1ddff8487f88251d118cbb5a5130184385a65f8cR39-R47)

- Fixes #XXXX (GitHub issue number)
- Fixes CAL-XXXX (Linear issue number - should be visible at the bottom of the GitHub issue description)

## Mandatory Tasks (DO NOT REMOVE)

- [ ] I have self-reviewed the code (A decent size PR without self-review might be rejected).
- [ ] I have updated the developer docs in /docs if this PR makes changes that would require a [documentation change](https://cal.com/docs). If N/A, write N/A here and check the checkbox.
- [ ] I confirm automated tests are in place that prove my fix is effective or that my feature works.

## How should this be tested?

- Are there environment variables that should be set?
- What are the minimal test data to have?
- What is expected (happy path) to have (input and output)?
- Any other important info that could help to test that PR

## Checklist

- I haven't read the [contributing guide](https://github.com/calcom/cal.com/blob/main/CONTRIBUTING.md)
- My code doesn't follow the style guidelines of this project
- I haven't commented my code, particularly in hard-to-understand areas
- I haven't checked if my changes generate no new warnings

### CODE REVIEW

File: packages/prisma/extensions/exclude-locked-users.ts [hunk 1]
```diff
@@ -25,6 +25,14 @@ export function excludeLockedUsersExtension() {
   });
 }
 
+function safeJSONStringify(x: any) {
```
Comment (zomars) [inline packages/prisma/extensions/exclude-locked-users.ts:28]: Just being extra careful that invalid JSON doesn't block any queries.

File: packages/prisma/extensions/exclude-locked-users.ts [hunk 2]
```diff
@@ -35,8 +43,10 @@ async function excludeLockedUsers(
   query: <T>(args: T) => Promise<unknown>
 ) {
   args.where = args.where || {};
+  const whereString = safeJSONStringify(args.where);
+  const shouldIncludeLocked = whereString.includes('"locked":false');
   // Unless explicitly specified, we exclude locked users
-  if (args.where.locked === undefined) {
+  if (!shouldIncludeLocked) {
```
Comment (zomars) [inline packages/prisma/extensions/exclude-locked-users.ts:49]: This only checked for "locked" at top level in the where property. The new way includes nested AND and OR conditions.

**Review by emrysal — DISMISSED** [review]

File: packages/prisma/extensions/exclude-locked-users.ts [hunk 3]
```diff
@@ -35,8 +43,10 @@ async function excludeLockedUsers(
   query: <T>(args: T) => Promise<unknown>
 ) {
   args.where = args.where || {};
+  const whereString = safeJSONStringify(args.where);
+  const shouldIncludeLocked = whereString.includes('"locked":false');
```
Comment (keithwillcode) [inline packages/prisma/extensions/exclude-locked-users.ts:47]: Based on this var name, shouldn’t this be checking for locked: true?

Comment (zomars) [reply packages/prisma/extensions/exclude-locked-users.ts:47]: Actually we should only check for the property name on ensure is not undefined.
```suggestion
  const shouldIncludeLocked = whereString.includes('"locked":');
```

**Review by emrysal — APPROVED** [review]

**Pull request merged**